Initial metrics tile

llamb-jump opened the pull request:
This PR adds a new metrics tile which aggregates measurements from other tiles and publishes them to https://metrics.solana.com. There is a single metrics being published in this PR, `__test_seq`, in `src/disco/mux/fd_mux.c`.

This functionality can be viewed at https://metrics.solana.com:3000/d/b74ee3d7-cc7e-467d-95f0-c7f1c07a1ea0/new-dashboard?orgId=1&from=1695851580220&to=1695884020369. Credentials to log in and view the dashboard are in the Firedancer/IT shared 1password vault.

We publish using [InfluxDB line protocol](https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/).

Highlights / Interesting Points:
- Metrics uses OpenSSL because `metrics.solana.com` requires `https`. The OpenSSL configuration had to be changed in `deps.sh` to give us required functionality.
- The sandbox had to be relaxed for the metrics tile since it opens and closes sockets, uses OpenSSL, and requires DNS. The sandbox for the metrics tile allows files to be opened, does not use a network namespace, and does not use `setup_mountns()`. In a future PR we will handle DNS ourselves by crafting requests and parsing responses which will allow us to use `setup_mountns()` and greatly reduce the number of syscalls which are whitelisted.
- A basic metrics api available in `src/util/metrics/fd_metrics.h` with a more advanced option available in `src/util/metrics/fd_metrics_impl.h`.
- A minor bug was addressed in `src/app/fddev/txn.c` so the flag `--payload-base64-encoded` works now.

Follow on work:
- Additional metrics and reworking the metrics API depending on how it's used.
- Metrics testing in CI. Will likely involve spinning up a simple HTTPS server locally with Python

mmcgee-jump commented on src/app/fdctl/run/tiles/dedup.c:
```diff
@@ -12,6 +13,7 @@ init( fd_tile_args_t * args ) {
   /* calling fd_tempo_tick_per_ns requires nanosleep, it is cached with
      a FD_ONCE */
   fd_tempo_tick_per_ns( NULL );
+  metrics_boot( args->metrics_pod, metrics_verify, args->tile_idx );
```
Howcome this is happening in init? Might be poorly named but init is for privileged setup, and couldn't see anything privileged in there.

llamb-jump replied:
Updated to be in `run()`

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+}
+
+static void
+run( fd_tile_args_t * args ) {
```
Could we make this tile use mux like the other ones, and align it to the same style / semantics

llamb-jump replied:
The metrics tile won't use the mux tile for now since it does not require flow control and it's already implemented.

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+
+  /* Receive and print the response */
+  char response[4096];
+  ret = SSL_read( ssl, response, sizeof( response ) );
```
This is a blocking read? I think we want this to be async so it doesn't block the housekeeping loop. Producers could easily get backpressured or drop packets while waiting for this.

mmcgee-jump replied:
The lifecycle here is a little unclear to me. Do you need to call SSL_shutdown to send close notify? As well, one SSL_read might not return all the bytes? I think you need to keep calling read until you can at least parse the HTTP response for a 200, but you maybe just want to read the whole thing out.

llamb-jump replied:
This has all been reworked to use nonblocking OpenSSL.

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+    FD_LOG_NOTICE(( "response %s", response ));
+  }
+
+  /* Clean up */
```
Probably want to parse the response for a HTTP 200?

mmcgee-jump replied:
(and keep the buffered metrics in that case)

llamb-jump replied:
Updated to parse the response and log if it is not correct.

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+  }
+
+  /* Create an SSL context */
+  SSL_CTX * ctx = SSL_CTX_new( TLS_client_method() );
```
These kinds of things can be initialized once at tile boot

llamb-jump replied:
Updated.

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+  }
+
+  /* Resolve the server address */
+  struct hostent * server = gethostbyname( _server );
```
Per discussion, kind of hate this glibc function, god knows what it ends up doing. Could we just include the MUSL `gethostbyname` code directly?

mmcgee-jump replied:
I also think gethostbyname is blocking, which has the same issues as below. This could be async / polled as part of the run loop instead.

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+  }
+}
+
+static long allow_syscalls[] = {
```
I know it's a pain but we can we document why each of these are needed, same as the other tiles? It's very useful for auditing and code review.

llamb-jump replied:
Updated

mmcgee-jump commented on src/app/fdctl/config.c:
```diff
@@ -537,6 +558,42 @@ init_workspaces( config_t * config ) {
+  config->shmem.workspaces[ idx ].num_pages = 1;
+  idx++;
+
+  config->shmem.workspaces[ idx ].kind      = wksp_metrics_verify;
```
Curious why one workspace per tile -> metrics? My feeling is it might be cleaner with just one <everyone> -> metrics workspace

llamb-jump replied:
We discussed this a few weeks ago and opted for one workspace per tile type for security.

mmcgee-jump commented on src/app/fdctl/config/default.toml:
```diff
@@ -514,6 +514,15 @@ dynamic_port_range = "8000-10000"
     [tiles.bank]
         receive_buffer_size = 128
 
+    [tiles.metrics]
```
Please document heavily in line with the rest of the configuration

llamb-jump replied:
Updated

mmcgee-jump commented on src/app/fdctl/run/tiles/metrics.c:
```diff
@@ -0,0 +1,392 @@
+
+static void
+ssl_init( void ) {
+  SSL_library_init();
```
Are these all necessary during privileged boot? Can you add some comments. Also error checking, same as https://github.com/firedancer-io/firedancer/blob/b0f14a28f9a4f387ab35c481760d31a8c24cf3e1/src/app/fdctl/run/tiles/quic.c#L33

llamb-jump replied:
They are - none of them have a return value.

mmcgee-jump commented on src/util/metrics/fd_metrics.h:
```diff
@@ -0,0 +1,79 @@
+} Datapoint;
+
+void
+metrics_boot( uchar const * pod, const metrics_tile_t tile, const ulong idx );
```
The code in util/ has much more requirements than the rest of the codebase, since it's a library.

Could we follow the style in fd_log.h a lot more closely? Eg, these exported functions and types should be prefixed with `fd_` so importers don't get symbol collisions. Use FD_PROTOTYPES_BEGIN, I don't think fd_metrics_format should be exposed, or the enum etc?

As far as I can tell, fd_metrics.c could have an API that's basically like,

* fd_metrics_boot()`
* #FD_METRICS_ULONG(tag, value) ... (or fd_metrics_push).

llamb-jump replied:
Move out of `util/` and now the prototype now looks like
```
void
fd_metrics_push( uint tag, ulong value )

mmcgee-jump commented on src/util/metrics/fd_metrics.h:
```diff
@@ -0,0 +1,79 @@
+
+typedef struct {
+  metrics_datapoint_t type;
+  char                name[METRICS_NAME_SZ];
```
Don't think these strings should be passing between tiles, since it's a lot of copying. Could be cleaner to have tag be a u16, then you can eliminate the dcache completely and just send mcache lines which are one cache line. Then metrics tile would map the tags back to strings before posting.

llamb-jump committed:
Initial metrics tile https://github.com/firedancer-io/firedancer/pull/707

- Adds a basic metrics tile which aggregates measurements from other
tiles and posts them using InfluxDB line protocol over HTTPS to
metrics.solana.com.

- Includes a single metrics, __test_seq, which is incremented in
fd_mux.c

- The metrics tile uses OpenSSL to post using https. More features had
to be enabled in deps.sh to allow posting using https.

- The sandbox has been relaxed for the metrics tile to allow the
creation of sockets.
